spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,109 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.datasketches.memory.Memory;
+import org.apache.datasketches.theta.Sketch;
+import org.apache.datasketches.theta.Sketches;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.spark.SparkReadOptions;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.stats.ThetaSketchAggregator;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  public static final String APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY = "ndv";
+
+  static List<Blob> generateNDVSketchesAndBlobs(
+      SparkSession spark, Table table, Snapshot snapshot, Set<String> columns) {
+    Map<Integer, Sketch> columnToSketchMap = computeNDVSketches(spark, table, snapshot, columns);
+    return generateBlobs(table, columns, columnToSketchMap, snapshot);
+  }
+
+  private static List<Blob> generateBlobs(
+      Table table, Set<String> columns, Map<Integer, Sketch> sketchMap, Snapshot snapshot) {
+    return columns.stream()
+        .map(
+            columnName -> {
+              Schema schema = table.schemas().get(snapshot.schemaId());
+              Types.NestedField field = schema.findField(columnName);
+              Sketch sketch = sketchMap.get(field.fieldId());
+              long ndv = (long) sketch.getEstimate();
+              return new Blob(
+                  StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                  ImmutableList.of(field.fieldId()),
+                  snapshot.snapshotId(),
+                  snapshot.sequenceNumber(),
+                  ByteBuffer.wrap(sketch.toByteArray()),
+                  null,
+                  ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, String.valueOf(ndv)));
+            })
+        .collect(Collectors.toList());
+  }
+
+  private static Map<Integer, Sketch> computeNDVSketches(
+      SparkSession spark, Table table, Snapshot snapshot, Set<String> columnsToBeAnalyzed) {
+    Map<Integer, Sketch> sketchMap = Maps.newHashMap();
+    String tableName = table.name();
+    List<String> columns = ImmutableList.copyOf(columnsToBeAnalyzed);
```

Comment:
I not a big a fan of using both `Set` and `List`, it gets confusing quickly. Let's accept `List` in this class and update the action to dedup the values.